forta-bots/No-Settle/src/agent.ts

```diff
@@ -17,60 +18,125 @@ export const provideHandleBlock = (
 
-    //Get past events on first run
+    //Initialize the domains & get past events on first run
     if (!init) {
       init = true;
-      const filter = {
+
+      const numDomains: BigNumber = await fetcher.getNumDomains(data.TeleportRouter, blockEvent.block.number);
+      for (let i: number = 0; i < numDomains.toNumber(); i++) {
+        domains.push(await fetcher.getDomain(data.TeleportRouter, i, blockEvent.block.number));
+      }
+
+      domains.forEach((domain) => {
+        latestAlertedSettleTimestampMap.set(domain, BigNumber.from(-1));
+        latestSettleTimestampMap.set(domain, BigNumber.from(0));
+      });
+
+      const settleFilter = {
         address: data.TeleportJoin,
         topics: [SETTLE_IFACE.getEventTopic("Settle")],
         fromBlock: blockEvent.block.number - daysThreshold * 6100, // 6100 blocks/day on average
         toBlock: blockEvent.block.number - 1,
       };
 
-      const settleEvents: providers.Log[] = await provider.getLogs(filter);
+      const settleEvents: providers.Log[] = await provider.getLogs(settleFilter);
 
       if (settleEvents.length) {
-        latestSettleTimestamp = BigNumber.from(
-          (await provider.getBlock(settleEvents[settleEvents.length - 1].blockNumber)).timestamp
+        await Promise.all(
+          domains.map(async (domain) => {
+            const domainSettleEvents: providers.Log[] = settleEvents.filter(
+              (event: providers.Log) => event.topics[1] === domain
+            );
+            if (domainSettleEvents.length) {
+              latestSettleTimestampMap.set(
+                domain,
+                BigNumber.from(
+                  (await provider.getBlock(domainSettleEvents[domainSettleEvents.length - 1].blockNumber)).timestamp
+                )
+              );
+            }
+          })
         );
       }
     }
 
-    // Check if:
+    const fileFilter = {
+      address: data.TeleportRouter,
+      topics: [FILE_IFACE.getEventTopic("File")],
+      blockHash: blockEvent.blockHash,
+    };
+
+    const fileEvents: providers.Log[] = await provider.getLogs(fileFilter);
+
+    if (fileEvents.length) {
+      domains.length = 0;
+      const numDomains: BigNumber = await fetcher.getNumDomains(data.TeleportRouter, blockEvent.block.number);
+      for (let i: number = 0; i < numDomains.toNumber(); i++) {
+        const domain = await fetcher.getDomain(data.TeleportRouter, i, blockEvent.block.number);
+        domains.push(domain);
+      }
+      domains.forEach((domain) => {
+        if (!latestSettleTimestampMap.has(domain)) {
+          latestAlertedSettleTimestampMap.set(domain, BigNumber.from(-1));
+          latestSettleTimestampMap.set(domain, BigNumber.from(0));
+        }
+      });
+      for (const key in latestSettleTimestampMap.keys()) {
+        if (!domains.includes(key)) {
+          latestAlertedSettleTimestampMap.delete(key);
+          latestSettleTimestampMap.delete(key);
+        }
+      }
```

Comment:
Efficiency improvement suggestion
```suggestion
      let newAlertedSettleTimestampMap: Map<string, BigNumber> = new Map<string, BigNumber>();
      let newSettleTimestampMap: Map<string, BigNumber> = new Map<string, BigNumber>();
      domains.forEach((domain) => {
        if (!latestSettleTimestampMap.has(domain)) {
          newAlertedSettleTimestampMap.set(domain, BigNumber.from(-1));
          newSettleTimestampMap.set(domain, BigNumber.from(0));
        }
        else {
          newAlertedSettleTimestampMap.set(domain, latestAlertedSettleTimestampMap.get(domain));
          newSettleTimestampMap.set(domain, latestSettleTimestampMap.get(domain));
        }
      });
      latestAlertedSettleTimestampMap = newAlertedSettleTimestampMap;
      latestSettleTimestampMap = newSettleTimestampMap;
```